okta products for devops post (#1133)
oktadeveloper/okta-blog: _source/_posts/2022-04-07-where-okta-meets-devops.md

---
layout: blog_post
title: "How Can DevOps Engineers Use Okta?"
author: edunham 
by: advocate
communities: [devops]

> description: "It's more than just SSO"

Reviewer: ```suggestion
description: "It's more than just SSO. This overview describes using Okta to access SaaS services and tooling, automate administration tasks, manage server access, and more."
This description is excerpted in a host of places, so more detail is useful for SEO. 

```

tags: [devops]
tweets:

> - ""

Author: ```suggestion
- "Do you know the 4 main ways that DevOps professionals can benefit from Okta's products?"
```

> - ""

Author: ```suggestion
- "Could an identity provider help manage your CI/CD systems, SSH access controls, and more?"
```

> - ""

Reviewer 2: Add tweets 🐦

> - ""

Author: ```suggestion
- "DevOps doesn't work without managing identity. What tools will you choose?"
```

image: blog/where-okta-meets-devops/devops.png
type: awareness
---

## What is DevOps? 

> DevOps is one of those buzzwords. Ask three DevOps professionals what it means, and you might get four different answers. Most will agree that DevOps usually describes a mix of tools and techniques combining aspects of systems administration, software development, IT, release engineering, security, and often QA. At its best, DevOps synergizes the strengths of these fields to reduce toil through automation and improve software quality through continuous testing and deployment, and [infrastructure as code](https://en.wikipedia.org/wiki/Infrastructure_as_code). At its worst, DevOps is used as an excuse to ask a single under-resourced team to do the work of several groups, in an ever-increasing spiral of tech debt. Most DevOps roles fall somewhere on the spectrum between these extremes. 

Reviewer 2: I love this:
> At its worst, DevOps is used as an excuse to ask a single under-resourced team to do the work of several groups, in an ever-increasing spiral of tech debt

While many tech professionals know Okta and its competitors as the way they access their corporate email and chat, there are actually four main areas where Okta's product family interfaces with the DevOps skillset. 

## Accessing SaaS services and tooling

> In 2022, it's vanishingly rare for a tech company to host every software product that they use on- premises. Although there are open or [self-hosted versions](https://github.com/awesome-selfhosted/awesome-selfhosted) of every major tool in the modern workflow, maintaining separate services gets prohibitively expensive. Instead of hosting physical servers, most organizations rent cloud resources. Instead of running email ((Mailman)[https://list.org/]), chat ([IRC](https://datatracker.ietf.org/doc/html/rfc1459), [XMPP](https://xmpp.org/about/technology-overview/)), user account management ([LDAP](https://ldap.com/)), and more in-house, companies instead rent services providing these features from specialist providers. 

Reviewer 2: ```suggestion
In 2022, it's vanishingly rare for a tech company to host every software product that they use on- premises. Although there are open or [self-hosted versions](https://github.com/awesome-selfhosted/awesome-selfhosted) of every major tool in the modern workflow, maintaining separate services gets prohibitively expensive. Instead of hosting physical servers, most organizations rent cloud resources. Instead of running email ([Mailman](https://list.org/)), chat ([IRC](https://datatracker.ietf.org/doc/html/rfc1459), [XMPP](https://xmpp.org/about/technology-overview/)), user account management ([LDAP](https://ldap.com/)), and more in-house, companies instead rent services providing these features from specialist providers. 
```

As well as the obvious workplace services like Slack and email, DevOps engineers might use their Okta accounts to sign in to cloud service providers like [AWS](https://www.okta.com/aws/), [GCP](https://www.okta.com/%20partners/google/), or [Azure](https://www.okta.com/integrations/azure-portal-login/#overview).

In a CI/CD pipeline, one might use Okta to log in to [Jenkins](https://www.okta.com/integrations/jenkins/#overview), [Spinnaker](https://docs.armory.io/armory-enterprise/armory-admin/auth-okta-configure/), or even [CircleCI](https://www.okta.com/integrations/circleci/#overview). If an organization uses [GitHub Enterprise](https://saml-doc.okta.com/Provisioning_Docs/GitHub_Enterprise_Provisioning.html), DevOps personnel can use Okta to sign in there too. 

## Automating Okta administration tasks
Just as sysadmins and IT managed identity for small to medium organizations in the olden days, DevOps engineers may be responsible for some administrative tasks in their company's Okta instance. 

> Software and infrastructure administration and management tasks can be automated using a plethora of tools, including Hashicorp's Terraform. Terraform allows users to describe the desired state of their resources, and then changes the resources to match that description.  Using Okta's [Terraform provider](https://github.com/okta/terraform-provider-okta), users can manage their Okta resources and track their changes with version control. This is particularly useful for those whose business needs have them managing many different Okta accounts and trying to keep the settings consistent across all of them. 

Reviewer 2: ```suggestion
Software and infrastructure administration and management tasks can be automated using a plethora of tools, including Hashicorp's Terraform. Terraform allows users to describe the desired state of their resources, and then changes the resources to match that description.  Using [Okta's Terraform provider](https://github.com/okta/terraform-provider-okta), users can manage their Okta resources and track their changes with version control. This is particularly useful for those whose business needs have them managing many different Okta accounts and trying to keep the settings consistent across all of them. 
```

> Software and infrastructure administration and management tasks can be automated using a plethora of tools, including Hashicorp's Terraform. Terraform allows users to describe the desired state of their resources, and then changes the resources to match that description.  Using [Okta's Terraform provider](https://github.com/okta/terraform-provider-okta), users can manage their Okta resources and track their changes with version control. This is particularly useful for those whose business needs have them managing many different Okta accounts and trying to keep the settings consistent across all of them. 

Reviewer 3: can okta be used to manage access to TF cloud?

Author: good catch, it looks like we can!

Author: I've added that to the discussion of using okta to login to SaaS, though, as answering "how does a customer manage access to modules which it publishes" right next to the question "how does a customer benefit from using the tf that okta publishes" seems like it could get confusing

## Managing Server Access

Modern code and infrastructure best practices recommend architecting systems to require minimal human access and intervention. But production-readiness is always a balancing act between aspirations and real-world limitations, and many successful codebases were architected before modern tooling made these hands-off design patterns feasible. Updating production code and infrastructure is often likened to repairing a jet engine while in flight: it has to keep working after every change that you make. When logging into servers is an important part of existing workflows, streamlining that process can be an essential step toward moving past the need to fix things by hand at all. 

> One way to improve the security and user experience of SSHing or RDPing to servers is to stop asking users to manage their own keys, and stop asking servers to keep track of who's allowed to access them. In an organization whose identity provider already tracks which team members should have infrastructure access, Okta's [ASA](https://www.okta.com/demo/okta-advanced-server-access/) (formerly ScaleFT) provides a great way to reduce the value of SSH and RDP credentials by making them ephemeral. 

Reviewer 2: ```suggestion
One way to improve the security and user experience of SSHing or RDPing to servers is to stop asking users to manage their own keys, and stop asking servers to keep track of who's allowed to access them. In an organization whose identity provider already tracks which team members should have infrastructure access, [Okta's ASA](https://www.okta.com/demo/okta-advanced-server-access/) (formerly ScaleFT) provides a great way to reduce the value of SSH and RDP credentials by making them ephemeral. 
```

## In Conclusion

Although the field itself is ambiguously defined, DevOps tools and practices constantly interface with identity. DevOps done right ensures that the people can access the resources they should, and not the ones they shouldn't. DevOps practitioners can use Okta to control identity-related behavior of their infrastructure through its many integrations, and can in turn customize and manage their Okta instances by using Terraform and Okta's metrics and monitoring APIs. 

If you work in DevOps, you can't escape the responsibility of managing identity as it relates to your infrastructure. But you can choose how to manage it: Will your organization support you in delegating these tasks to specialists, do you have a use case where it's better to attempt to reinvent these wheels yourself?

> * For more ways to watch what Okta is doing from within your own tools, see the [developer docs on monitoring](https://developer.okta.com/docs/concepts/monitor). 
> * For more about the Okta Terraform provider, check out the [GitHub repo](https://github.com/okta/terraform-provider-okta)
> * To learn about securing Kubernetes with OpenID Connect and Okta, read [this blog post](/blog/2021/11/08/k8s-api-server-oidc)
> * For further information about managing server access with Okta's ASA, see [ASA's overview docs](https://help.okta.com/asa/en-us/Content/Topics/Adv_Server_Access/docs/asa-overview.htm)

Reviewer 2: At one point someone suggested I used better names for link text for SEO reasons.  I have no idea if that's true or not, so... insert a grain of🧂 below

```suggestion
* For more ways to watch what Okta is doing from within your own tools, see the [developer docs on monitoring](https://developer.okta.com/docs/concepts/monitor). 
* Checkout out the [Okta Terraform provider](https://github.com/okta/terraform-provider-okta)
* To learn about [securing Kubernetes with OpenID Connect and Okta](/blog/2021/11/08/k8s-api-server-oidc)
* For further information about managing server access with Okta's ASA, see [ASA's overview docs](https://help.okta.com/asa/en-us/Content/Topics/Adv_Server_Access/docs/asa-overview.htm)
```

---
layout: blog_post
title: "How Can DevOps Engineers Use Okta?"
author: edunham 
by: advocate
communities: [devops]

> description: "It's more than just SSO. This overview describes using Okta to access SaaS services and tooling, automate administration tasks, manage server access, and more."

Reviewer: ```suggestion
description: "It's more than just SSO. This overview describes using Okta to access SaaS services and tooling, automate administration tasks, manage server access, and more."
```
This description is excerpted in a host of places, so more detail is useful for SEO.

> This description is excerpted in a host of places, so more detail is useful for SEO. 

Reviewer: No idea why this turned up as a standalone line/ addition to front matter, I thought I'd added it as a comment for why I edited the description. Sorry, doesn't belong here
```suggestion

```

image: blog/where-okta-meets-devops/devops.png
type: awareness
---

## What is DevOps? 

DevOps is one of those buzzwords. Ask three DevOps professionals what it means, and you might get four different answers. Most will agree that DevOps usually describes a mix of tools and techniques combining aspects of systems administration, software development, IT, release engineering, security, and often QA. At its best, DevOps synergizes the strengths of these fields to reduce toil through automation and improve software quality through continuous testing and deployment, and [infrastructure as code](https://en.wikipedia.org/wiki/Infrastructure_as_code). At its worst, DevOps is used as an excuse to ask a single under-resourced team to do the work of several groups, in an ever-increasing spiral of tech debt. Most DevOps roles fall somewhere on the spectrum between these extremes. 

While many tech professionals know Okta and its competitors as the way they access their corporate email and chat, there are actually four main areas where Okta's product family interfaces with the DevOps skillset. 

## Accessing SaaS services and tooling

> In 2022, it's vanishingly rare for a tech company to host every software product that they use on- premises. Although there are open or [self-hosted versions](https://github.com/awesome-selfhosted/awesome-selfhosted) of every major tool in the modern workflow, maintaining separate services gets prohibitively expensive. Instead of hosting physical servers, most organizations rent cloud resources. Instead of running email ([Mailman](https://list.org/)), chat ([IRC](https://datatracker.ietf.org/doc/html/rfc1459), [XMPP](https://xmpp.org/about/technology-overview/)), user account management ([LDAP](https://ldap.com/)), and more in-house, companies instead rent services providing these features from specialist providers. 

Reviewer: ```suggestion
In 2022, it's vanishingly rare for a tech company to host every software product that they use on-premises. Although there are open or [self-hosted versions](https://github.com/awesome-selfhosted/awesome-selfhosted) of every major tool in the modern workflow, maintaining separate services gets prohibitively expensive. Instead of hosting physical servers, most organizations rent cloud resources. Instead of running email ([Mailman](https://list.org/)), chat ([IRC](https://datatracker.ietf.org/doc/html/rfc1459), [XMPP](https://xmpp.org/about/technology-overview/)), user account management ([LDAP](https://ldap.com/)), and more in-house, companies instead rent services providing these features from specialist providers. 
```
extra space

Before the pattern of using a central identity provider and protocols like OAuth became commonplace, users would have to manage a separate login and password to every software-as-a-service vendor that their company used. This created security and administrative headaches for everyone involved. But thanks to widespread adoption of the underlying protocols, a company can now use a single identity provider like Okta to allow everyone in the organization to access all of the software that it buys for them. 

As well as the obvious workplace services like Slack and email, DevOps engineers might use their Okta accounts to sign in to cloud service providers like [AWS](https://www.okta.com/aws/), [GCP](https://www.okta.com/%20partners/google/), or [Azure](https://www.okta.com/integrations/azure-portal-login/#overview).

> In a CI/CD pipeline, one might use Okta to log in to [Jenkins](https://www.okta.com/integrations/jenkins/#overview), [Spinnaker](https://docs.armory.io/armory-enterprise/armory-admin/auth-okta-configure/), or even [CircleCI](https://www.okta.com/integrations/circleci/#overview). If an organization uses [GitHub Enterprise](https://saml-doc.okta.com/Provisioning_Docs/GitHub_Enterprise_Provisioning.html), DevOps personnel can use Okta to sign in there too. 

Author: ```suggestion
In a CI/CD pipeline, one might use Okta to log in to [Jenkins](https://www.okta.com/integrations/jenkins/#overview), [Spinnaker](https://docs.armory.io/armory-enterprise/armory-admin/auth-okta-configure/), or even [CircleCI](https://www.okta.com/integrations/circleci/#overview). If an organization uses [GitHub Enterprise](https://saml-doc.okta.com/Provisioning_Docs/GitHub_Enterprise_Provisioning.html), DevOps personnel can use Okta to sign in there too. Okta can even help ensure that the right people can update infrastructure code by managing access to [Docker Hub](https://www.okta.com/integrations/docker-hub/#overview) or [Terraform Cloud](https://www.okta.com/integrations/terraform-cloud/#overview)!
```

## Automating Okta administration tasks
Just as sysadmins and IT managed identity for small to medium organizations in the olden days, DevOps engineers may be responsible for some administrative tasks in their company's Okta instance. 

> Software and infrastructure administration and management tasks can be automated using a plethora of tools, including Hashicorp's Terraform. Terraform allows users to describe the desired state of their resources, and then changes the resources to match that description.  Using [Okta's Terraform provider](https://github.com/okta/terraform-provider-okta), users can manage their Okta resources and track their changes with version control. This is particularly useful for those whose business needs have them managing many different Okta accounts and trying to keep the settings consistent across all of them. 

Reviewer: ```suggestion
Software and infrastructure administration and management tasks can be automated using a plethora of tools, including HashiCorp's Terraform. Terraform allows users to describe the desired state of their resources, and then changes the resources to match that description.  Using [Okta's Terraform provider](https://github.com/okta/terraform-provider-okta), users can manage their Okta resources and track their changes with version control. This is particularly useful for those whose business needs have them managing many different Okta accounts and trying to keep the settings consistent across all of them. 
```
camelcase of HashiCorp - company name

> ## Managing Server Access

Reviewer: ```suggestion
## Managing server access
```
unless it's referring to a product name, the heading should be sentence case

One way to improve the security and user experience of SSHing or RDPing to servers is to stop asking users to manage their own keys, and stop asking servers to keep track of who's allowed to access them. In an organization whose identity provider already tracks which team members should have infrastructure access, [Okta's ASA](https://www.okta.com/demo/okta-advanced-server-access/) (formerly ScaleFT) provides a great way to reduce the value of SSH and RDP credentials by making them ephemeral. 

Additionally, Okta is working on [more new products](https://www.okta.com/press-room/press-releases/okta-introduces-new-okta-privileged-access-product-to-strengthen-security/) for privileged access management, which will be interesting in the DevOps use case. 

## Introspecting Okta

When DevOps refers to a catch-all role focused on the care and feeding of infrastructure, monitoring and alerting fall under its purview. 

> So far, we've talked about information flowing from Okta to your software and services, answering questions like "is this login request legitimate?". Automating Okta administration tasks with Terraform pushes configuration from the user to Okta. There's one more direction that information can move relative to your identity provider, however: a stream of metadata about what the identity provider is doing can flow out of the provider and into your monitoring and alerting tools. 

Reviewer: ```suggestion
So far, we've talked about information flowing from Okta to your software and services, answering questions like "is this login request legitimate?" Automating Okta administration tasks with Terraform pushes configuration from the user to Okta. There's one more direction that information can move relative to your identity provider, however: a stream of metadata about what the identity provider is doing can flow out of the provider and into your monitoring and alerting tools. 
```
removed period, per: https://docs.microsoft.com/en-us/style-guide/punctuation/quotation-marks (see examples)

## In Conclusion

Although the field itself is ambiguously defined, DevOps tools and practices constantly interface with identity. DevOps done right ensures that the people can access the resources they should, and not the ones they shouldn't. DevOps practitioners can use Okta to control identity-related behavior of their infrastructure through its many integrations, and can in turn customize and manage their Okta instances by using Terraform and Okta's metrics and monitoring APIs. 

> If you work in DevOps, you can't escape the responsibility of managing identity as it relates to your infrastructure. But you can choose how to manage it: Will your organization support you in delegating these tasks to specialists, do you have a use case where it's better to attempt to reinvent these wheels yourself?

Reviewer: ```suggestion
If you work in DevOps, you can't escape the responsibility of managing identity as it relates to your infrastructure. But you can choose how to manage it: Will your organization support you in delegating these tasks to specialists? Do you have a use case where it's better to attempt to reinvent these wheels yourself?
```
This is an important idea, a key summary of your argument. I'd recommend expanding it into 2 separate questions, giving the reader more room to absorb it... Also possible to add "Or" at the opening: "Or do you have a use case..."
